### PR TITLE
Sql storage

### PR DESCRIPTION
**Pull request by MaksyKun**
- Added SQL support for the whole storage (exp, professions, queues, auto-crafting, etc.)
- Added auto-migration into sql on plugin initialization
- Added auto-backup of old storage files (data.yml & players-folder) into fusion/backup
- Added a fix for the BrowseGUI were newly added professions after /craft reload couldnt be learned
- Added tab-completition for the commands (with permission checks too)

### Note:
- There is one "bug" so far that annoyed me in the development about the visual remaining time of QueueItems. At the moment, some queue-items get stuck on one seconds which show a time of [18 -> 17 -> 16 -> **16** -> 15 -> 14, ...]. Its an acceptable bug, but Id be glad if we could manage to fix this together someday.
- This PR makes breaking changes since the whole FileConfiguration-Handling of data.yml and players-folder are replaced by SQLite, MySQL or MariaDB. Even though there is a backup method, I recommend to still backup data manually in case you test on production server data.

### CODE REVIEW

**Comment by MaksyKun** [comment]
Gonna fix few things that i realized afterwards before its ready again.

**Comment by Travja** [comment]
I've just kind of skimmed so far. Does this effectively remove yaml-based configs? Then we're doing storage locally with sqlite or with a proper mysql database?

**Comment by MaksyKun** [comment]
Yes exactly. The yaml-based files for each players data will be changed to sqlite (default) or mysql/mariadb if you wish to. It should improve performance as well as multi server support.

**Comment by MaksyKun** [comment]
Is there any improvement required from my side yet, @Travja?

File: pom.xml [hunk 1]
```diff
@@ -72,5 +72,15 @@
...
+
+        <dependency>
+            <groupId>org.xerial</groupId>
+            <artifactId>sqlite-jdbc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.mysql</groupId>
+            <artifactId>mysql-connector-j</artifactId>
+            <version>${mysql.version}</version>
+        </dependency>
```
Comment (Travja) [inline pom.xml:84]: I went ahead and removed the mariadb dependency. I don't think it's necessary with the mysql-connector (which I also updated to use the version provided by the parent project).

**Review by Travja — COMMENTED** [review]
I think things look good. Let me know if there is anything you're concerned about at all.

**Review by Travja — APPROVED** [review]

File: src/main/java/studio/magemonkey/fusion/Commands.java [hunk 2]
```diff
@@ -166,6 +172,31 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
                     MessageUtil.sendMessage("fusion.help", sender, new MessageData("sender", sender),
                             new MessageData("text", label + " " + StringUtils.join(args, ' ')));
                 }
+            } else if(args[0].equalsIgnoreCase("storage")) {
+                String storage = args[1];
+                DatabaseType type = DatabaseType.valueOf(Cfg.getConfig().getString("storage.type", "LOCALE").toUpperCase());
+                switch (storage.toLowerCase()) {
+                    case "locale":
```
[the remark this answers is not in the document]
Comment (MaksyKun) [reply src/main/java/studio/magemonkey/fusion/Commands.java]: I have renamed all things from locale -> local.

File: pom.xml (hunk 1 above)
Comment (MaksyKun) [reply pom.xml:84]: Do you think I should remove the MariaDBConnection in generel or keep it for anything in futute?
image

Comment (Travja) [reply pom.xml:84]: MariaDB and MySQL *should* be able to operate off the same MySQL driver, I believe. So it's fine to stay. Our other plugins don't have explicit support for maria db though

**Comment by Travja** [comment]
Looks like you'll also need to merge in `dev` since it looks like there are some merge conflicts

**Comment by MaksyKun** [comment]
> Looks like you'll also need to merge in `dev` since it looks like there are some merge conflicts

Meaning:
1.  I merge sql-storage into my local dev
2. Resolve the conflicts
3. Push dev?

**Comment by Travja** [comment]
`git checkout dev`
`git pull`
`git checkout sql-storage`
`git merge dev`
resolve conflicts... 
`git push`

File: pom.xml (hunk 1 above)
Comment (MaksyKun) [reply pom.xml:84]: Alrighty

**Comment by MaksyKun** [comment]
I just found some stuff regarding to categoryPattern parameter that needs to be fixed. Ill put the PR in draft for the meantime.

**Comment by MaksyKun** [comment]
Looks fine now

**Pull request merged**